### ISSUE TITLE
[code] Does not work with alpine 

### ISSUE THREAD

**Issue by csweichel** [gitpod-io/gitpod#3356]
Gitpod Code does not work with alpine images because it requires libc-based native modules. The hack we've used for Theia doesn't seem to work here.

We should decide if this is something we want to support going forward, and if so, how we'll fix it.

See also https://github.com/gitpod-io/gitpod/issues/3355

If we wanted to support this, maybe installing glibc within the userland might do the trick.

**Comment by AlexTugarev** [comment]
> The hack we've used for Theia doesn't seem to work here.

did we already tried?

**Comment by meysholdt** [comment]
here is the "[hack](https://github.com/gitpod-io/gitpod/blob/main/components/theia/app/package-libs.sh)": It's a script that moves all native libraries into a folder and re-links them so that they become a self-contained unit. Thus code would bundle the needed libc-based native modules. The script needs to identify all native binaries... maybe that needs some adjustment for code?

What I like about this approach is that it solves the compatibility issue for *all* linux distros and not just Alpine. It allowed us to *not* have distro-specific Theia variants. 

However, if vs code extensions contain native binaries they would either need to support the underlying linux distro or be processed by this script, too.

**Comment by aledbf** [comment]
Not sure exactly what's the problem. Adding the package `libc6-compat` allows execution of binaries that require libc.
Test repository with theia works https://github.com/aledbf/gitpod-musl

**Comment by gtsiolis** [comment]
A community member reported this, too. See [relevant discussion](https://community.gitpod.io/t/workspace-doesnt-start-with-custom-docker-image/3806).

**Comment by dnsmichi** [comment]
I ran into the same problem when using Gitpod as a live review app within GitLab, and Hugo for Pages. The default Hugo image uses Alpine, and it took me some days to comment out the right portions in the .gitpod.yml to figure out that the error is the image, and not the ports.

https://twitter.com/dnsmichi/status/1413524263865393157?s=21

https://gitlab.com/everyonecancontribute/web/everyonecancontribute.gitlab.io/-/commit/9e7da3d2a7e539b188796c581d60879d64df5b2d

The builder just times out, no error message, no logs, no insights. I'd suggest improving this in 2 ways:

1. Give the error messages / console log to the user when the build timed out
2. Parse the .gitpod.yml before building (you do that already), and inject a custom rule into the config parser, which detects `.*alpine.*` for the `image` key. Decide whether error - break or warning - log, and give the user a hindsight to this issue, including a note that Alpine does not work at the moment.

From there, the first impression helps with adoptions and switching to a Debian based image, and potentially working together with linking some libc magic other projects do as well. 

Thanks @gtsiolis for help and support 🙏

**Comment by dnsmichi** [comment]
Forgot to mention - the Alpine image was added when I was using Theia since 2020. I switched to Code last week before we started learning Rust in Gitpod again - https://everyonecancontribute.com/post/2021-06-30-cafe-36-rust-rocket-prometheus/ I came across this problem while writing a blogpost for the GitLab blog, trying to use Hugo as example for review apps.

**Comment by blakewest** [comment]
Hi, I ran into this issue as well. I switched back to Theia, and the gitpod space is working again. I'm wondering is there a list of supported images that can be used with Code? If so, it would be awesome to list them here, so that perhaps an easy switch can be made in the docker container. Or some other migration instructions.

**Comment by akosyakov** [comment]
@blakewest You can find VS Code base requirements here: https://code.visualstudio.com/docs/remote/linux#_tips-by-linux-distribution

**Comment by a16bitsysop** [comment]
The alpine wiki says gcompat is better:
https://wiki.alpinelinux.org/wiki/Running_glibc_programs

https://git.adelielinux.org/adelie/gcompat

But that did not work for me either, maybe it needs symlinking to another path?

https://pkgs.alpinelinux.org/contents?branch=edge&name=gcompat&arch=x86_64&repo=community

**Comment by etienne1911** [comment]
Do you know if this issue affects other distributions such as arch linux or is only related to alpine?

**Comment by a16bitsysop** [comment]
Arch Linux is glibc based not musl so it should work, I tried to create a custom image from debian-slim and it errored out when integrating the Gitpod layer during launch, so I used gitpod/workspace-base with an alpine chroot which works now apart from making a single command/script to get into the chroot as Gitpod user.

The issue seems to be after you make a custom image Gitpod adds it’s own layer to make it work with Gitpod, which is based on glibc and possibly has other expectations?

**Comment by etienne1911** [comment]
I tried to launch archiso as docker image, but also had issue with gitpod layer..
I suppose it requires to replicate some steps from workspace-base... to set new gitpod user notably..
Currently have another issue with pacman which was reported before and probably caused by  glibc version on older docker version.

On the other hand, I looked at setting an arch chroot inside a default container. and after several attempts, I finally managed to make it work

**Comment by a16bitsysop** [comment]
I have created a custom gitpod docker image now for alpine that starts in the alpine chroot, with the alpine abuild tools installed:
https://gitlab.com/a16bitsysop/alpine_gitpod

**Comment by TheBeeBoi** [comment]
Now I know why this wasn't working on my dockerimage, I thought it had something to do with the user..

**Comment by gbraad** [comment]
Got a base image working with the following: https://github.com/gbraad-devenv/alpine/commit/07db75b4957234cd3e18ad1bded19f4b131aed05.

Full repo at: https://github.com/gbraad-devenv/alpine/
